src/py/mat3ra/made/tools/build/nanoribbon/builders.py

```diff
@@ -0,0 +1,121 @@
+from .configuration import NanoribbonConfiguration
+from ...modify import center_material
+
+
+class NanoribbonBuilder(BaseBuilder):
+    _ConfigurationType: type(NanoribbonConfiguration) = NanoribbonConfiguration  # type: ignore
+    _GeneratedItemType: Material = Material
+    _PostProcessParametersType: Any = None
+
+    @staticmethod
+    def _calculate_dimensions(config: NanoribbonConfiguration, material: Material):
+        provided_width = config.width
+        provided_length = config.length
+        vacuum_width = config.vacuum_width
+        vacuum_length = config.vacuum_length
+        edge_type = config.edge_type
+
+        if edge_type == "armchair":
+            provided_length, provided_width = provided_width, provided_length
+            vacuum_width, vacuum_length = vacuum_length, vacuum_width
+
+        length = provided_length * np.dot(np.array(material.basis.cell.vector1), np.array([1, 0, 0]))
+        width = provided_width * np.dot(np.array(material.basis.cell.vector2), np.array([0, 1, 0]))
+        height = np.dot(np.array(material.basis.cell.vector3), np.array([0, 0, 1]))
+        vacuum_length = vacuum_length * np.dot(np.array(material.basis.cell.vector1), np.array([1, 0, 0]))
+        vacuum_width = vacuum_width * np.dot(np.array(material.basis.cell.vector2), np.array([0, 1, 0]))
+
+        return length, width, height, vacuum_length, vacuum_width
+
+    @staticmethod
+    def _get_new_lattice_vectors(length, width, height, vacuum_length, vacuum_width, edge_type):
+        length_lattice_vector = [length + vacuum_length, 0, 0]
+        width_lattice_vector = [0, width + vacuum_width, 0]
+        height_lattice_vector = [0, 0, height]
+
+        if edge_type == "armchair":
+            length_lattice_vector, width_lattice_vector = width_lattice_vector, length_lattice_vector
+
+        return length_lattice_vector, width_lattice_vector, height_lattice_vector
+
+    def _create_supercell_and_cut_nanoribbon(
+        self, material, provided_length, provided_width, length, width, height, edge_type
+    ):
+        supercell_size_coeff = 1
+        edge_nudge_value = 0.01
+        conditional_nudge_value = edge_nudge_value * (-1 * (edge_type == "armchair") + 1 * (edge_type == "zigzag"))
+
+        if edge_type == "armchair":
+            supercell_size_coeff = supercell_size_coeff * (provided_width // provided_length + 1)
+
+        min_coordinate = [-edge_nudge_value, conditional_nudge_value, 0]
+        max_coordinate = [length - edge_nudge_value, width + conditional_nudge_value, height]
+        supercell = create_supercell(
+            material, np.diag([2 * provided_length * supercell_size_coeff, 2 * provided_width, 1])
+        )
+        nanoribbon = filter_by_rectangle_projection(
+            supercell, min_coordinate=min_coordinate, max_coordinate=max_coordinate, use_cartesian_coordinates=True
+        )
+
+        return nanoribbon
+
+    @staticmethod
+    def _update_basis_and_lattice(
```

Comment:
Move to a generic place - Material, modify

Reply:
Moving to Material